### DISCUSSION TITLE
Faster or customizable git submodule checkout

### DISCUSSION THREAD

**Discussion by Kobzol** [renovatebot/renovate#33027] [Suggest an Idea]
### Tell us more.

Hi! I'm a member of the Rust infra team, and I'd like to get renovatebot working on the https://github.com/rust-lang/rust repository, which hosts the Rust toolchain (compiler, standard library etc.), using the Forking Renovatebot GitHub app.

Currently, I'm facing two blockers:
- We need to use nightly cargo (this is already tracked in https://github.com/renovatebot/renovate/discussions/15728 and other issues)
- Slow speed of git submodules checkout

For updating certain Rust lockfiles, we need to checkout submodules in the Rust repository. However, the repo has a lot of submodules, some of which are very large (e.g. LLVM or GCC). I tried to set `"cloneSubmodules": true`, but then the renovatebot job times out after thirty minutes.

I wonder if it would be possible to either:
- Allow selecting which submodules will be checked out. We actually need only a few small-ish submodules for the lockfile updates, but currently Renovatebot either checks out all submodules or none.
- Make submodule checkouts faster, by using some git magic. For example, in the Rust repo, we have a [custom script](https://github.com/rust-lang/rust/blob/04acc1d9dcb57b973e818ea817b106af5214164f/src/ci/scripts/checkout-submodules.sh) that makes the checkouts faster by downloading archives from GitHub (although this is not a general solution, of course).

Thanks!

**Comment by rarkins** [comment]
Hi, we'd be really interested in helping you get the rust repo working well with Renovate Bot!

The logic for checking out submodules iterates through them all, so in theory could have a filter available to filter in or out some of them: https://github.com/renovatebot/renovate/blob/05c13eb7a1cf8276daa94ce1b9c4b3bdf1759f78/lib/util/git/index.ts#L347-L367

Alternatively, we have migrated our backend to MicroVMs recently which allow us to be more flexible with "script" enablement (it's no longer such a security concern). We planned to make features like these available to OSS projects, so running your repo's script could be part of a "post clone commands" option we add to Renovate.

Which of these two would be faster, if both options were available?

**Reply by Kobzol** [reply]
If we wanted to check out all submodules, then a custom script would be faster. However, for our current needs, we only need a few relatively small submodules (we need Rust git submodules for updating lockfiles, we don't need the gargantuan C/C++ dependency submodules for that), and it seems like the ability to checkout selected submodules could be useful to more people. E.g. the current `cloneSubmodules` boolean could be generalized to an array of submodules that would be checked out. That would allow us to only checkout the submodules that we need.

**Comment by rarkins** [comment]
Code for this is here: https://github.com/renovatebot/renovate/pull/33115

I haven't tested it on a real repo yet, so it's not ready to merge though.

**Comment by Kobzol** [comment]
Thank you! We're using the Renovatebot GitHub app, so with that it's probably not possible to test your branch, but I suppose that I could self-host it and point it at our repo (not sure how difficult that is)?

**Reply by rarkins** [reply]
I'll try testing first. Which filters would you apply? Ie include only certain ones or exclude certain ones?

**Reply by Kobzol** [reply]
That's a good question :) Since it seemed like we had the most problems with checking out the gcc submodule, I would probably try disabling `src/gcc` and also `src/llvm-project` first. Disabling large submodules is easier to maintain than to explicitly enumerate all the "fast" submodules, which might be changing in the future.

**Reply by Kobzol** [reply]
https://developer.mend.io/github/Kobzol/rust/-/job/b74c5ce8-4407-4f61-94c6-b2e8081612aa is the time outed submodule checkout, performed on my [fork](https://github.com/kobzol/rust) of the Rust repo. Not sure if you can access the job though.

**Reply by rarkins** [reply]
I'll reproduce and test locally using that repo, thanks

**Comment by rarkins** [comment]
Converted to issue: https://github.com/renovatebot/renovate/issues/33157
